apps/api_web/lib/api_web/rate_limiter/rate_limiter_concurrent.ex

```diff
@@ -0,0 +1,175 @@
+defmodule ApiWeb.RateLimiter.RateLimiterConcurrent do
+  @moduledoc """
+  Rate limits a user's concurrent connections based on their API key or by their IP address if no
+  API key is provided. Split by static and event-stream requests.
+  """
+
+  use GenServer
+  require Logger
+
+  def start_link([]), do: GenServer.start_link(__MODULE__, nil, name: __MODULE__)
+
+  def init(_) do
+    connection_opts =
+      Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :connection_opts)
+
+    {:ok, pid} = if memcache?(), do: Memcache.start_link(connection_opts), else: {:ok, nil}
+    {:ok, %{memcache_pid: pid}}
+  end
+
+  defp lookup(%ApiWeb.User{} = user, event_stream?) do
+    type = if event_stream?, do: "event_stream", else: "static"
+    key = "concurrent_#{user.id}_#{type}"
+
+    {
+      type,
+      key
+    }
+  end
+
+  defp get_current_unix_ts do
+    System.system_time(:second)
+  end
+
+  defp get_heartbeat_tolerance do
+    Keyword.fetch!(Application.get_env(:api_web, :rate_limiter_concurrent), :heartbeat_tolerance)
+  end
+
+  def get_locks(%ApiWeb.User{} = user, event_stream?) do
+    if enabled?() do
+      current_timestamp = get_current_unix_ts()
+      heartbeat_tolerance = get_heartbeat_tolerance()
+      {_type, key} = lookup(user, event_stream?)
+      {:ok, locks} = GenServer.call(__MODULE__, {:memcache_get, key, %{}})
+      # Check if any expired, and remove:
+      valid_locks =
+        :maps.filter(
+          fn _, timestamp ->
+            timestamp + heartbeat_tolerance >= current_timestamp
+          end,
+          locks
+        )
+
+      if valid_locks != locks, do: GenServer.call(__MODULE__, {:memcache_set, key, valid_locks})
+      valid_locks
+    else
+      %{}
+    end
+  end
+
+  def check_concurrent_rate_limit(user, event_stream?) do
+    active_connections = user |> get_locks(event_stream?) |> Map.keys() |> length
+
+    limit =
+      case {event_stream?, user.type} do
+        {true, :registered} ->
+          if user.streaming_concurrent_limit >= 0,
+            do:
+              max(
+                user.streaming_concurrent_limit || 0,
+                Keyword.fetch!(
```

Comment:
suggestion: can we do these application lookups at compile time instead, since they aren't changing at runtime?